<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Docutils 0.5: http://docutils.sourceforge.net/" />
<title>Seqdb2 Documentation</title>
<meta name="author" content="Alex Nolley" />
<meta name="date" content="February 9, 2009" />
<style type="text/css">

/*
:Author: David Goodger (goodger@python.org)
:Id: $Id: html4css1.css 5196 2007-06-03 20:25:28Z wiemann $
:Copyright: This stylesheet has been placed in the public domain.

Default cascading style sheet for the HTML output of Docutils.

See http://docutils.sf.net/docs/howto/html-stylesheets.html for how to
customize this style sheet.
*/

/* used to remove borders from tables and images */
.borderless, table.borderless td, table.borderless th {
  border: 0 }

table.borderless td, table.borderless th {
  /* Override padding for "table.docutils td" with "! important".
     The right padding separates the table cells. */
  padding: 0 0.5em 0 0 ! important }

.first {
  /* Override more specific margin styles with "! important". */
  margin-top: 0 ! important }

.last, .with-subtitle {
  margin-bottom: 0 ! important }

.hidden {
  display: none }

a.toc-backref {
  text-decoration: none ;
  color: black }

blockquote.epigraph {
  margin: 2em 5em ; }

dl.docutils dd {
  margin-bottom: 0.5em }

/* Uncomment (and remove this text!) to get bold-faced definition list terms
dl.docutils dt {
  font-weight: bold }
*/

div.abstract {
  margin: 2em 5em }

div.abstract p.topic-title {
  font-weight: bold ;
  text-align: center }

div.admonition, div.attention, div.caution, div.danger, div.error,
div.hint, div.important, div.note, div.tip, div.warning {
  margin: 2em ;
  border: medium outset ;
  padding: 1em }

div.admonition p.admonition-title, div.hint p.admonition-title,
div.important p.admonition-title, div.note p.admonition-title,
div.tip p.admonition-title {
  font-weight: bold ;
  font-family: sans-serif }

div.attention p.admonition-title, div.caution p.admonition-title,
div.danger p.admonition-title, div.error p.admonition-title,
div.warning p.admonition-title {
  color: red ;
  font-weight: bold ;
  font-family: sans-serif }

/* Uncomment (and remove this text!) to get reduced vertical space in
   compound paragraphs.
div.compound .compound-first, div.compound .compound-middle {
  margin-bottom: 0.5em }

div.compound .compound-last, div.compound .compound-middle {
  margin-top: 0.5em }
*/

div.dedication {
  margin: 2em 5em ;
  text-align: center ;
  font-style: italic }

div.dedication p.topic-title {
  font-weight: bold ;
  font-style: normal }

div.figure {
  margin-left: 2em ;
  margin-right: 2em }

div.footer, div.header {
  clear: both;
  font-size: smaller }

div.line-block {
  display: block ;
  margin-top: 1em ;
  margin-bottom: 1em }

div.line-block div.line-block {
  margin-top: 0 ;
  margin-bottom: 0 ;
  margin-left: 1.5em }

div.sidebar {
  margin: 0 0 0.5em 1em ;
  border: medium outset ;
  padding: 1em ;
  background-color: #ffffee ;
  width: 40% ;
  float: right ;
  clear: right }

div.sidebar p.rubric {
  font-family: sans-serif ;
  font-size: medium }

div.system-messages {
  margin: 5em }

div.system-messages h1 {
  color: red }

div.system-message {
  border: medium outset ;
  padding: 1em }

div.system-message p.system-message-title {
  color: red ;
  font-weight: bold }

div.topic {
  margin: 2em }

h1.section-subtitle, h2.section-subtitle, h3.section-subtitle,
h4.section-subtitle, h5.section-subtitle, h6.section-subtitle {
  margin-top: 0.4em }

h1.title {
  text-align: center }

h2.subtitle {
  text-align: center }

hr.docutils {
  width: 75% }

img.align-left {
  clear: left }

img.align-right {
  clear: right }

ol.simple, ul.simple {
  margin-bottom: 1em }

ol.arabic {
  list-style: decimal }

ol.loweralpha {
  list-style: lower-alpha }

ol.upperalpha {
  list-style: upper-alpha }

ol.lowerroman {
  list-style: lower-roman }

ol.upperroman {
  list-style: upper-roman }

p.attribution {
  text-align: right ;
  margin-left: 50% }

p.caption {
  font-style: italic }

p.credits {
  font-style: italic ;
  font-size: smaller }

p.label {
  white-space: nowrap }

p.rubric {
  font-weight: bold ;
  font-size: larger ;
  color: maroon ;
  text-align: center }

p.sidebar-title {
  font-family: sans-serif ;
  font-weight: bold ;
  font-size: larger }

p.sidebar-subtitle {
  font-family: sans-serif ;
  font-weight: bold }

p.topic-title {
  font-weight: bold }

pre.address {
  margin-bottom: 0 ;
  margin-top: 0 ;
  font-family: serif ;
  font-size: 100% }

pre.literal-block, pre.doctest-block {
  margin-left: 2em ;
  margin-right: 2em }

span.classifier {
  font-family: sans-serif ;
  font-style: oblique }

span.classifier-delimiter {
  font-family: sans-serif ;
  font-weight: bold }

span.interpreted {
  font-family: sans-serif }

span.option {
  white-space: nowrap }

span.pre {
  white-space: pre }

span.problematic {
  color: red }

span.section-subtitle {
  /* font-size relative to parent (h1..h6 element) */
  font-size: 80% }

table.citation {
  border-left: solid 1px gray;
  margin-left: 1px }

table.docinfo {
  margin: 2em 4em }

table.docutils {
  margin-top: 0.5em ;
  margin-bottom: 0.5em }

table.footnote {
  border-left: solid 1px black;
  margin-left: 1px }

table.docutils td, table.docutils th,
table.docinfo td, table.docinfo th {
  padding-left: 0.5em ;
  padding-right: 0.5em ;
  vertical-align: top }

table.docutils th.field-name, table.docinfo th.docinfo-name {
  font-weight: bold ;
  text-align: left ;
  white-space: nowrap ;
  padding-left: 0 }

h1 tt.docutils, h2 tt.docutils, h3 tt.docutils,
h4 tt.docutils, h5 tt.docutils, h6 tt.docutils {
  font-size: 100% }

ul.auto-toc {
  list-style-type: none }

</style>
</head>
<body>
<div class="document" id="seqdb2-documentation">
<h1 class="title">Seqdb2 Documentation</h1>
<table class="docinfo" frame="void" rules="none">
<col class="docinfo-name" />
<col class="docinfo-content" />
<tbody valign="top">
<tr><th class="docinfo-name">Author:</th>
<td>Alex Nolley</td></tr>
<tr><th class="docinfo-name">Date:</th>
<td>February 9, 2009</td></tr>
</tbody>
</table>
<!-- contents: -->
<div class="section" id="introduction">
<h1>Introduction</h1>
<p>Seqdb2 is a database generation and querying package made to be used with gene
sequences generated by Solexa machines, namely FASTA and FASTQ formats. Values
such as sequence name, sequence description and the sequence itself can be
retrieved from these databases. seqdb2 consists of a c++ database writer and a
python/pyrex reader. Records from FASTA or FASTQ text files are parsed into a
database by the writer and then the records can be retrieved by index or name
with the reader.</p>
</div>
<div class="section" id="getting-going">
<h1>Getting Going</h1>
<p>The following software packages are required to build and run seqdb2:</p>
<ul class="simple">
<li>Python 2.4 or newer</li>
<li>gcc</li>
<li>make</li>
<li>cmake</li>
<li>pyrex</li>
</ul>
<p>If you want to run the included tests after building, you will need to install
nose as well</p>
<div class="section" id="downloading">
<h2>Downloading</h2>
<dl class="docutils">
<dt>You will need git to download a copy from the public git repository:</dt>
<dd>git-clone git://iorich.caltech.edu/git/public/seqdb2</dd>
</dl>
</div>
<div class="section" id="installing">
<h2>Installing</h2>
<p>Assuming you have already downloaded the package, this is how you build it:</p>
<pre class="literal-block">
cd seqdb2/
cmake .
make
</pre>
<p>To run the optional nosetests type:</p>
<pre class="literal-block">
nosetests
</pre>
</div>
</div>
<div class="section" id="quick-start">
<h1>Quick-Start</h1>
<div class="section" id="using-the-writers">
<h2>Using the Writers</h2>
<p>At the moment there is no Python wrapper for the C++ database builders, so
they are run from the command line. There is a writer for FASTA files and a
writer for FASTQ files named fadbm and fqdbm respectivly. Databases made from
either are completly accessible from the reader. The format for using the
database writers is:</p>
<pre class="literal-block">
./fadbm PATH_TO_FASTA_FILE
</pre>
<p>or:</p>
<pre class="literal-block">
./fqdbm PATH_TO_FASTQ_FILE.
</pre>
<p>So, if you wanted to create a database from the tests/test.fa file the code is:</p>
<pre class="literal-block">
bin/fadbm tests/test.fa
</pre>
<p>And now there are the files test.fa_seqdb2, test.fa_seqdb2_idx and
test.fa_seqdb2_hash under the files/ directory. These three files are the
database of test.fa containing the raw data, indicies of records and the hashed
names of each record.</p>
</div>
<div class="section" id="using-the-reader">
<h2>Using the Reader</h2>
<p>Now that the databases are generated. Querying them is simple and can be done
from the Pyrex module</p>
<p>In the python interpreter, import the seqdb2 module from the python
subdirectory and load some databases:</p>
<pre class="literal-block">
&gt;&gt;&gt; import seqdb2
&gt;&gt;&gt; fadb = seqdb2.dbread('../tests/test.fa_seqdb2')
&gt;&gt;&gt; fqdb = seqdb2.dbread('../tests/test.fastq_seqdb2')
</pre>
<p>The databases act like read-only dictionary objects:</p>
<pre class="literal-block">
&gt;&gt;&gt; fadb.keys()
&gt;&gt;&gt; fqdb.keys()
</pre>
<p>If you want to retrieve the names of the records from a database:</p>
<pre class="literal-block">
&gt;&gt;&gt; names = fadb.keys()
</pre>
<p>Or, to retrieve sequence information:</p>
<pre class="literal-block">
&gt;&gt;&gt; sequences = []
&gt;&gt;&gt; for key in fqdb:
...     record = fqdb[key]
...     sequences.append(record.sequence)
</pre>
<p>Length of the databases are easily found:</p>
<pre class="literal-block">
&gt;&gt;&gt; print len(fadb)
22
&gt;&gt;&gt; print len(fqdb)
49
</pre>
<p>Records can also be retrieved by index:</p>
<pre class="literal-block">
&gt;&gt;&gt; fqdb.loadRecordByIndex(5)
</pre>
<!-- Local Variables:
mode: rst
mode: outline-minor
End: -->
</div>
</div>
</div>
</body>
</html>
